src/ripple/protocol/TER.h

```diff
@@ -218,6 +292,135 @@ enum TER
     tecEXPIRED                  = 148
 };
 
+// Allow TECcodes to be promoted to TER.
+template <>
+class CanCvtToTER<TECcodes> : public std::true_type
+{ explicit CanCvtToTER() = delete; };
+
+//------------------------------------------------------------------------------
+
+// Template class that is specific to selected ranges of error codes.  The
+// Trait tells std::enable_if which ranges are allowed.
+template <template<typename> class Trait>
+class TERSubset
+{
+    TERUnderlyingType code_;
+
+public:
+    // Constructors
+    constexpr TERSubset() : code_ (tesSUCCESS) { }
+    constexpr TERSubset (TERSubset const& rhs) = default;
+    constexpr TERSubset (TERSubset&& rhs) = default;
+    constexpr explicit TERSubset (int rhs) : code_ (rhs) { } // For TxMeta
+
+    template <typename T, typename = std::enable_if_t<Trait<T>::value>>
+    constexpr TERSubset (T rhs) : code_ (static_cast<int>(rhs)) { }
+
+    // Assignment
+    constexpr TERSubset& operator=(TERSubset const& rhs) = default;
+    constexpr TERSubset& operator=(TERSubset&& rhs) = default;
+
+    template <typename T>
+    constexpr auto
+    operator= (T rhs) -> std::enable_if_t<Trait<T>::value, TERSubset&>
+    { code_ = static_cast<int>(rhs); return *this; }
+
+    // Conversion operators.
+    // Conversion to int is not explicit to allow streaming without casting.
+    constexpr operator int() const { return code_; }
```

Comment:
I started wondering how big of a pain it would be to make this conversion `explicit`, and thus a little safer.  It wasn't that bad: https://github.com/HowardHinnant/rippled/commit/fcd55a85f32470e1f72599850504ef77f54255b8

Reply:
Thanks for poking at that, @HowardHinnant.  That's a great thing to think about.  And it's cool that making the conversion operator explicit had basically no negative consequences.  You took the middle road and I took two (slightly) more radical roads.

1. @seelabs had noticed that if `TERSubset` has a conversion to `int` operator, then none of the comparison machines is needed.  So I made a branch that removes all of the comparison operators from `TERSubset`.  It worked just fine.  You can see the results here: https://github.com/scottschurr/rippled/commits/ter-check-no-cmp

2. For contrast, I thought it would be interesting to see if we could make it hard to accidentally convert from a `TER` to an `int`.  It turns out that the code base relies on being able to make those conversions.  But I thought they'd be easier to see if we gave names to those conversions.  So I tried it.  That branch also worked just fine.  You can see those results here: https://github.com/scottschurr/rippled/commits/ter-check-no-int

Reply:
I think I favor either named or explicit conversion to `int`. The implicit conversion to `int`, while simplifying the conversions, offers more potential for an accidental conversion leading to a run-time error.

Long-term, I'm currently leaning towards explicit conversions over named conversions as in general it reduces what I call the "apparent API".  I.e. one doesn't have to recall/lookup the name of the named conversion to convert from `T` to `U`.  It's always `U u{t};`  However that's long term, not short term.  gcc-5.2 won't even accept that syntax when `U` is `int`.  It's a bug that they fix somewhere in 6.x.  In the mean-time we have to limp along with `auto u = static_cast<U>(t);`.

Short-term, I'm fine with either named or explicit.